### PR TITLE
generic column (sample) key

### PR DESCRIPTION
**Pull request by cseed**
... and some infrastructure for generic row (variant) key.

Some VSM methods require sample key is String, basically those that use pedigrees.

I think it is feature complete but there is currently no way to create a non-string column key and there is no non-string column key tests.

### CODE REVIEW

File: python/hail/dataset.py [hunk 1]
```diff
@@ -19,13 +17,20 @@
 
 
 @decorator
+def requireSampleTString(func, vds, *args, **kwargs):
```
Comment (tpoterba) [inline python/hail/dataset.py]: shouldn't this go on the scala side? it doesn't do any conversion, which is why requireTGenotype is on the py side.

File: src/main/scala/is/hail/methods/KinshipMatrix.scala [hunk 2]
```diff
@@ -122,9 +124,9 @@ case class KinshipMatrix(val hc: HailContext, val matrix: IndexedRowMatrix, val
   def exportIdFile(idFile: String) {
     hc.sc.hadoopConfiguration.writeTextFile(idFile) { s =>
       for (id <- sampleIds) {
-        s.write(id)
+        s.write(id.asInstanceOf[String])
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: seems unsafe. Goes along with my comments about moving the schema checks to scala.

File: src/main/scala/is/hail/methods/LinearMixedRegression.scala [hunk 3]
```diff
@@ -39,7 +39,7 @@ object LinearMixedRegression {
 
     val (y, cov, completeSamples) = RegressionUtils.getPhenoCovCompleteSamples(assocVds, yExpr, covExpr)
     val completeSamplesSet = completeSamples.toSet
-    val sampleMask = assocVds.sampleIds.map(completeSamplesSet).toArray
+    val sampleMask = assocVds.sampleIds.map(s => completeSamplesSet(s)).toArray
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/LinearMixedRegression.scala]: do you prefer this style?

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 4]
```diff
@@ -204,6 +204,24 @@ object VariantSampleMatrix {
       case _ => false
     }
 
+    val sSignature = fields.get("sample_schema") match {
+      case Some(t: JString) => Parser.parseType(t.s)
+      case Some(other) => fatal(
+        s"""corrupt VDS: invalid metadata
+           |  Expected `JString' in field `genotype_schema', but got `${ other.getClass.getName }'
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: sample_schema

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 5]
```diff
@@ -204,6 +204,24 @@ object VariantSampleMatrix {
...
+           |  Recreate VDS with current version of Hail.""".stripMargin)
+      case _ => TString
+    }
+
+    val vSignature = fields.get("variant_schema") match {
+      case Some(t: JString) => Parser.parseType(t.s)
+      case Some(other) => fatal(
+        s"""corrupt VDS: invalid metadata
+           |  Expected `JString' in field `genotype_schema', but got `${ other.getClass.getName }'
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: sample_schema

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 6]
```diff
@@ -778,8 +747,8 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VSMMetadata,
       annotateSamples(m.getOrElse(_, nullValue), finalType, inserter)
     } else {
       keyTypes match {
-        case Array(TString) =>
-          var r = keyedRDD.map { case (k, v) => (k.asInstanceOf[Row].getAs[String](0), v: Annotation) }
+        case Array(`sSignature`) =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: I can't see the rest of the code easily but I think you need to fix the error message below

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 7]
```diff
@@ -906,7 +875,7 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VSMMetadata,
 
     } else {
       keyTypes match {
-        case Array(TVariant) =>
+        case Array(`vSignature`) =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: I can't see the rest of the code easily but I think you need to fix the error message below

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 8]
```diff
@@ -1240,16 +1209,16 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VSMMetadata,
...
-    val p = (s: String, sa: Annotation) => Filter.keepThis(samples.contains(s), keep)
+  def filterSamplesList(samples: Set[Annotation], keep: Boolean = true): VariantSampleMatrix[T] = {
+    val p = (s: Annotation, sa: Annotation) => Filter.keepThis(samples.contains(s), keep)
     filterSamples(p)
   }
 
   def filterSamplesTable(table: KeyTable, keep: Boolean): VariantSampleMatrix[T] = {
     table.keyFields.map(_.typ) match {
-      case Array(TString) =>
+      case Array(`sSignature`) =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: fix error message below

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 9]
```diff
@@ -1311,7 +1280,7 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VSMMetadata,
   def filterVariantsTable(kt: KeyTable, keep: Boolean = true): VariantSampleMatrix[T] = {
     val keyFields = kt.keyFields.map(_.typ)
     val filt = keyFields match {
-      case Array(TVariant) =>
+      case Array(`vSignature`) =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: fix error message below

File: src/test/scala/is/hail/methods/LinearRegressionSuite.scala [hunk 10]
```diff
@@ -266,7 +267,7 @@ class LinearRegressionSuite extends SparkSuite {
 
     val vds = hc.importVCF("src/test/resources/regressionLinear.vcf")
       .annotateSamplesTable(covariates, root = "sa.cov")
-      .annotateSamplesFam("src/test/resources/regressionLinear.fam")
```
Comment (tpoterba) [inline src/test/scala/is/hail/methods/LinearRegressionSuite.scala]: can you delete annotateSamplesFam from VSM?

Comment (cseed) [inline src/test/scala/is/hail/methods/LinearRegressionSuite.scala]: Didn't I?

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 11]
```diff
@@ -925,7 +928,7 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VSMMetadata,
         case other =>
           fatal(
             s"""method 'annotate_variants_table' expects a key table keyed by one of the following:
-               |  [ Variant ]
+               |  [ $vSignature ]
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: This is a little janky. OK for now though

Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: we basically need to build rules about which types can be promoted to other types

Comment (cseed) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: Why janky?  That's literally what the code above says:

 >         case Array(`vSignature`) =>

I agree, in the long term we could reasonably support conversions from the function registry in matching up keys on joins.

Comment (tpoterba) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: I mean variant is now a variable, but locus and Interval are still hard coded.

Comment (cseed) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: Ah.  Yes!  In my coming code, Locus is not hardcoded, but Interval still is.  I think if we make Interval type parameterized: Interval[T], then everything generalizes nicely.

**Pull request merged**